kernel/src/table_changes/resolve_dvs.rs

```diff
@@ -0,0 +1,422 @@
+            // `Treemap_s(2)` for deleted rows.
+            //
+            //  # Insertion Selection Treemap
+            //  The selection vector of added rows is calculated using set subtraction over deletion
+            //  treemaps `rm_dv - add_dv`. These rows went from set (deleted) in `rm_dv` to unset
+            //  (restored) in the `add_dv`. All other rows are either cancelled in set subtraction,
+            //  or were not selected in either treemap. Hence, they are not selected to
+            //  be in the set of changed rows. Applying this to our deletion treemaps:
+            //  rm_dv - add_dv =
+            //      Treemap_d(0, 1)
+            //    - Treemap_d(1, 2)
+            //    = Treemap_s(0)
+            //  The selection treemap shows that row 0 was inserted
+            //
+            //  # Deletion Selection Treemap
+            //  The selection vector of deleted rows is calculated using `add_dv - rm_dv`. These rows went
+            //  from unset (present) in `rm_dv` to set (deleted) in the `add_dv`. Once again, all
+            //  other rows are either cancelled in set subtraction, or were not set in either treemap.
+            //  Applying this to our deletion vectors:
+            //  add_dv - rm_dv =
+            //      Treemap_d(1, 2)
+            //    - Treemap_d(0, 1)
+            //    = Treemap_s(2)
+            //  The selection treemap shows that row 2 was deleted
+            //
+            //  # Conversion to Selection Vector
+            //  The selection treemap is converted to a selection vector by setting the bits:
+            //      Treemap_s(0) => [true]
+            //      Treemap_s(2) => [false, false, true]
+            //  All other rows are unselected (false).
+            let adds = &rm_dv - &add_dv;
+            let removes = add_dv - rm_dv;
+
+            let adds = (!adds.is_empty()).then_some(adds);
+            let removes = (!removes.is_empty()).then_some(removes);
+            (adds, removes)
+        }
+        (add_dv, None, CdfScanFileType::Add | CdfScanFileType::Cdc) => {
+            (Some(add_dv.unwrap_or_else(Default::default)), None)
+        }
+        (rm_dv, None, CdfScanFileType::Remove) => {
+            (None, Some(rm_dv.unwrap_or_else(Default::default)))
+        }
+    };
+    let treemap_to_bools = if scan_file.remove_dv.is_some() {
+        selection_treemap_to_bools
+    } else {
+        deletion_treemap_to_bools
+    };
+
+    let resolve = |scan_file, sv: Vec<bool>| ResolvedCdfScanFile {
+        scan_file,
+        selection_vector: (!sv.is_empty()).then_some(sv),
+    };
+
+    let removes = rm_dv.map(treemap_to_bools).map(|sv| {
+        let scan_file = CdfScanFile {
+            scan_type: CdfScanFileType::Remove,
+            ..scan_file.clone()
+        };
+        resolve(scan_file, sv)
+    });
+    let adds = add_dv
+        .map(treemap_to_bools)
+        .map(|sv| resolve(scan_file, sv));
+    Ok([removes, adds].into_iter().flatten())
```

Comment:
wonder if the iterator is the right approach? we only ever return one or two items, correct? does the (future) callsite necessitate the iterator?

Reply:
You're right that we only return one or two items.

The callsite doesn't necessitate it, but it [works nicely](https://github.com/delta-io/delta-kernel-rs/pull/580/files/852ad1c52a5b67cc951636162304e69b911c30ca..72decb32d79bff6aebc80826c01d0e8abfedc5f8#diff-9f259259e81e25dd6b4545ade21b50f737bac75da34c3db3edd50880930ada11R221) with flattening:
```
            .map(move |scan_file| {
                resolve_scan_file_dv(dv_engine_ref.as_ref(), &table_root, scan_file?)
            }) // Iterator-Result-Iterator

```

If you know of a better way tho, lmk 👀